Security solution mki prod testing

dkirchan opened the pull request:
## Summary

This PR adds the functionality to the existing testing framework for `Security Solution Serverless` to execute the tests in production in the provided Zone. 

So right now the combination of :
- process.env.ENVIRONMENT -> Can only be production, otherwise it defaults in QA. 
- process.env.REGION -> Defaults to 'aws-eu-west-1'

Gives the oppotunity to provide the combinations required. 

The default flow regarding **production** is going directly into the Cloud through the [cloud_project_handler](https://github.com/elastic/kibana/pull/189193/files#diff-17074621cd8a3ee40e5865fecec0a7ddf153997ccbfe58d8f33fa7d405dfb594) and not through the proxy handler to use multi org which for now works **ONLY FOR** QA environment.

dkirchan committed:
Fixed Cypress/API tests running against prod project

dkirchan committed:
Fixed pipelines

dkirchan committed:
Fixed genai tests

dkirchan committed:
Omitted BK analytics key as it is exported above

dkirchan committed:
Fixed region env param

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

dkirchan commented:
/ci

dkirchan committed:
Merge branch 'main' into security-solution-mki-prod-testing

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

Ikuni17 reviewed (commented):
The OPs code lgtm, but I'm unsure why this needs to run against prod? Shouldn't the quality gates run beforehand?

cc @elastic/appex-qa for visibility

maximpn commented on x-pack/plugins/security_solution/scripts/run_cypress/parallel_serverless.ts:
```diff
@@ -58,12 +58,12 @@ const PROVIDERS = Object.freeze({
     const jsonString = fs.readFileSync(path.join(userHomeDir, '.elastic/cloud.json'), 'utf-8');
     const jsonData = JSON.parse(jsonString);
-    return jsonData.api_key.qa;
+    return environment === 'prod' ? jsonData.api_key.prod : jsonData.api_key.qa;
```
nit: What values `environment` can have?

It looks beneficial to define enum or constants for known environments. Something like

```ts
enum Environment {
  Production = 'prod',
  QA = 'qa'
}
```

and use it instead of hardcoded values. WDYT?

dkirchan replied:
Sounds better. On it.

maximpn commented on x-pack/plugins/security_solution/scripts/run_cypress/parallel_serverless.ts:
```diff
@@ -247,36 +247,47 @@ const getProductTypes = (
 export const cli = () => {
   run(
     async (context) => {
+      const environment = process.env.ENVIRONMENT ? `${process.env.ENVIRONMENT}` : 'qa';
```
```suggestion
      const environment = process.env.ENVIRONMENT ?? : 'qa';
```

maximpn commented on x-pack/plugins/security_solution/scripts/run_cypress/parallel_serverless.ts:
```diff
@@ -247,36 +247,47 @@ const getProductTypes = (
     async (context) => {
+      const environment = process.env.ENVIRONMENT ? `${process.env.ENVIRONMENT}` : 'qa';
+
+      const PROXY_URL = process.env.PROXY_URL ? process.env.PROXY_URL : undefined;
```
```suggestion
      const PROXY_URL = process.env.PROXY_URL;
```

maximpn commented on x-pack/plugins/security_solution/scripts/run_cypress/parallel_serverless.ts:
```diff
@@ -247,36 +247,47 @@ const getProductTypes = (
+      const environment = process.env.ENVIRONMENT ? `${process.env.ENVIRONMENT}` : 'qa';
+
+      const PROXY_URL = process.env.PROXY_URL ? process.env.PROXY_URL : undefined;
+      const PROXY_SECRET = process.env.PROXY_SECRET || undefined;
```
```suggestion
      const PROXY_SECRET = process.env.PROXY_SECRET;
```

`process.env.PROXY_SECRET`'s type is `string | undefined`. Do we expect something else here?

maximpn commented on x-pack/plugins/security_solution/scripts/run_cypress/parallel_serverless.ts:
```diff
@@ -247,36 +247,47 @@ const getProductTypes = (
+
+      const PROXY_URL = process.env.PROXY_URL ? process.env.PROXY_URL : undefined;
+      const PROXY_SECRET = process.env.PROXY_SECRET || undefined;
+      const PROXY_CLIENT_ID = process.env.PROXY_CLIENT_ID || undefined;
```
```suggestion
      const PROXY_CLIENT_ID = process.env.PROXY_CLIENT_ID;
```

maximpn commented on x-pack/plugins/security_solution/scripts/run_cypress/parallel_serverless.ts:
```diff
@@ -247,36 +247,47 @@ const getProductTypes = (
       // This works for either local executions or fallback in case proxy service is unavailable.
-      if (!process.env.CLOUD_QA_API_KEY && !getApiKeyFromElasticCloudJsonFile()) {
-        log.error('The API key for the environment needs to be provided with the env var API_KEY.');
+      if (!process.env.CLOUD_API_KEY && !getApiKeyFromElasticCloudJsonFile(environment)) {
```
nit: `getApiKeyFromElasticCloudJsonFile(environment)` could be invoked once and the returned value reused twice instead of calling `getApiKeyFromElasticCloudJsonFile(environment)` twice.

dkirchan replied:
Fixed.

maximpn commented on x-pack/plugins/security_solution/scripts/run_cypress/parallel_serverless.ts:
```diff
@@ -247,36 +247,47 @@ const getProductTypes = (
+        PROXY_URL &&
+        PROXY_CLIENT_ID &&
+        PROXY_SECRET &&
+        environment.toUpperCase() !== 'prod'.toUpperCase() &&
```
Why casing normalization is only used here? Don't we need to do the same in `getApiKeyFromElasticCloudJsonFile()`? I'd recommend normalize `environment`'s value at the beginning on the line `250`. It will help to match case insensitively. And it should match with my comment above regarding known `environment` values.



`environment.toUpperCase() !== 'prod'.toUpperCase()` can be statically transformed to `environment.toUpperCase() !== 'PROD'`. Though why casing is only used here but

dkirchan replied:
Agreed!

maximpn commented on x-pack/plugins/security_solution/scripts/run_cypress/project_handler/proxy_project_handler.ts:
```diff
@@ -31,9 +31,10 @@ export class ProxyHandler extends ProjectHandler {
     productTypes?: ProductType[],
     commit?: string
   ): Promise<Project | undefined> {
+    const region = process.env.REGION ? process.env.REGION : DEFAULT_REGION;
```
nit: There is interesting chain of names `process.env.REGION` -> `region` -> `region_id`. It means that `process.env.REGION` is a region id. Maybe use `regionId` as an intermediate constant name?

maximpn commented on x-pack/test/security_solution_api_integration/scripts/mki_api_ftr_execution.ts:
```diff
@@ -88,23 +89,48 @@ export const cli = () => {
+        PROXY_URL &&
+        PROXY_CLIENT_ID &&
+        PROXY_SECRET &&
+        environment.toUpperCase() !== 'prod'.toUpperCase() &&
```
The same as above regarding casing (`.toUpperCase()`).

maximpn commented on x-pack/test/security_solution_api_integration/scripts/mki_api_ftr_execution.ts:
```diff
@@ -88,23 +89,48 @@ export const cli = () => {
+
+      // Checking if API key is either provided via env variable or in ~/.elastic.cloud.json
+      // This works for either local executions or fallback in case proxy service is unavailable.
+      if (!process.env.CLOUD_API_KEY && !getApiKeyFromElasticCloudJsonFile(environment)) {
```
The same as above regarding double `getApiKeyFromElasticCloudJsonFile(environment)` invocation.

maximpn reviewed (commented):
@dkirchan Thanks for extending MKI testing functionality 👍 

I have the same question as @Ikuni17. How is it gonna work? Does it mean we will create MKI production environments and run tests against them? So that environments will be used only for tests. It also means that we'll deploy not yet released versions like BCs.

dkirchan commented:
@Ikuni17 this is not part of the quality gates. We were asked by the PMs to execute our tests once (on demand in the future) against new regions for the serverless projects. 

It will not be running on a scheduled basis and potentially this will be needed just a few more times.

dkirchan commented:
@maximpn As I mentioned in my previous comment it is needed only once/twice in order to validate safe functionality in different project regions. 

This will not be part of the teams process, quality gates or anything relevant. It just extends the functionality to create a project in production (with whichever image already exists there) and then run some tests against this project. 

It works similarly to the quality gate in QA but is completely irrelevant. It is sole responsibility of our team at least for now.

dkirchan committed:
Update x-pack/plugins/security_solution/scripts/run_cypress/parallel_serverless.ts

Co-authored-by: Maxim Palenov <maxim.palenov@elastic.co>

dkirchan committed:
Update x-pack/plugins/security_solution/scripts/run_cypress/parallel_serverless.ts

Co-authored-by: Maxim Palenov <maxim.palenov@elastic.co>

dkirchan committed:
Update x-pack/plugins/security_solution/scripts/run_cypress/parallel_serverless.ts

Co-authored-by: Maxim Palenov <maxim.palenov@elastic.co>

dkirchan committed:
Update x-pack/plugins/security_solution/scripts/run_cypress/parallel_serverless.ts

Co-authored-by: Maxim Palenov <maxim.palenov@elastic.co>

dkirchan committed:
Merge branch 'main' into security-solution-mki-prod-testing

maximpn commented:
Hi @dkirchan, thanks for the explanation 👍

Do you think we need to have the same explanation in the code? I suppose it may confuse code readers. Having explanation before `cli()` functions in `x-pack/test/security_solution_api_integration/scripts/mki_api_ftr_execution.ts` and `x-pack/plugins/security_solution/scripts/run_cypress/parallel_serverless.ts` will help to preserve this knowledge.

dkirchan committed:
Fixed comments for PR

dkirchan commented:
@maximpn I think that I addressed all the comments including the description around the prod testing above the respective CLIs. Testing right now and I would appreciate any updated comments you may have.

maximpn reviewed (approved):
@dkirchan thank you for addressing my comments 🙏 

I had a second look and left two minor comments but overall it looks good 👍

Ikuni17 commented:
@dkirchan I've talked with the QA team a bit about this PR and we'd like a bit more information on why this is required to run in prod and who was requesting that it run there. We would like to explore alternative solutions if possible.

@pheyos will be back from PTO next week and the QA team wanted to discuss the changes with him as well.

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
